File: Avalanchain.Cloud/Confirmation.fs
namespace Avalanchain.Cloud

module Confirmation = 
    open Avalanchain.Quorum

    type Confirmation<'T> = {
        NodeId: string
        ValueId: ValueId
        Value: 'T
        Notifier: 'T -> unit
    }
    and ValueId = string

    type ConfirmationResult<'T> =
        | InvalidConfirmation
        | ConfirmedSame
        | ConfirmedDifferent of 'T
        | NotConfirmedYet

    type ConfirmationCounter<'T when 'T: equality> (policy: ExecutionPolicy, validator, policyChecker) =
        let mutable confirmations = []
        let mutable invalidConfirmations = []
        let mutable pendingConfirmations = []
        let mutable confirmedValue = None
        member __.Policy = policy
        member __.AddConfirmation (confirmation: Confirmation<'T>) = 
            if not <| validator confirmation then 
                invalidConfirmations <- confirmation :: invalidConfirmations
                InvalidConfirmation
            else
                confirmations <- confirmation :: confirmations
                match confirmedValue with
                | Some v -> 
                    if confirmation.Value = v then ConfirmedSame
                    else ConfirmedDifferent v
                | None ->
                    confirmedValue <- policyChecker policy confirmations // TODO: Add possibility for reconfirmations
                    match confirmedValue with
                    | Some v -> 
                        for pc in pendingConfirmations do pc.Notifier v // Notifying pendings
                        if confirmation.Value = v then ConfirmedSame
                        else ConfirmedDifferent v
                    | None -> 
                        pendingConfirmations <- confirmation :: pendingConfirmations
                        NotConfirmedYet
        member __.Confirmations with get() = confirmations
        member __.InvalidConfirmations with get() = invalidConfirmations
        member __.PendingConfirmations with get() = pendingConfirmations
            
  
//    let ofQueue (queue: CloudQueue<'T>) f = 
//        asyncSeq { 
//            let centroidsSoFar = ResizeArray()
//            while true do
//                match queue.TryDequeue() with
//                | Some d ->                  
//                        yield d
//                        do! Async.Sleep 1
//                | None -> do! Async.Sleep 1
//        }
//        |> AsyncSeq.map(f)   

File: avalanchain-scala/backend/src/main/scala/com/avalanchain/core/confirmation.scala
package com.avalanchain.core

import com.avalanchain.core.domain.ChainStream.Hash
import com.avalanchain.core.domain.ExecutionPolicy

/**
  * Created by Yuriy Habarov on 19/04/2016.
  */
package object confirmation {
  type ValueId = String
  type ConfirmationStateChangedNotifier = Hash => Unit
  case class Confirmation (nodeId: String, valueId: ValueId, value: Hash, notifier: ConfirmationStateChangedNotifier)

  sealed trait ConfirmationResult {
    val confirmation: Confirmation
  }
  object ConfirmationResult {
    final case class InvalidConfirmation(confirmation: Confirmation) extends ConfirmationResult
    final case class ConfirmedSame(confirmation: Confirmation) extends ConfirmationResult
    final case class ConfirmedDifferent(confirmation: Confirmation, expectedHash: Hash) extends ConfirmationResult
    final case class NotConfirmedYet(confirmation: Confirmation) extends ConfirmationResult
  }

  type PolicyChecker = ExecutionPolicy => List[Confirmation] => Option[Hash]
  class ConfirmationCounter (val policy: ExecutionPolicy, validator: Confirmation => Boolean, policyChecker: PolicyChecker) {
    private var _confirmations = List.empty[Confirmation]
    private var _invalidConfirmations = List.empty[Confirmation]
    private var _pendingConfirmations = List.empty[Confirmation]
    private var _confirmedValue: Option[Hash] = None

    private def notifyDependents(): Unit = {
      _confirmedValue match {
        case Some(v) => _confirmations.foreach(c => c.notifier(v))
        case _ =>
      }
    }

    def confirmations = _confirmations
    def invalidConfirmations = _invalidConfirmations
    def pendingConfirmations = _pendingConfirmations
    def confirmedValue = _confirmedValue

    def addConfirmation(confirmation: Confirmation) = {
      if (!validator(confirmation)) {
        _invalidConfirmations = confirmation :: _invalidConfirmations
        ConfirmationResult.InvalidConfirmation(confirmation)
      }
      else {
        _confirmations = confirmation :: _confirmations
        _confirmedValue match {
          case Some(v) if v.equals(confirmation.value) => ConfirmationResult.ConfirmedSame(confirmation)
          case Some(v) => ConfirmationResult.ConfirmedDifferent(confirmation, v)
          case None => {
            _confirmedValue = policyChecker(policy) (_confirmations)
            _confirmedValue match {
              case Some(v) =>
                _confirmations = _pendingConfirmations
                _pendingConfirmations = List.empty[Confirmation]
                notifyDependents()
                if (v.equals(confirmation.value)) ConfirmationResult.ConfirmedSame(confirmation)
                else ConfirmationResult.ConfirmedDifferent(confirmation, v) // Shouldn't ever happen really because of policyChecker call
              case None =>
                _pendingConfirmations = confirmation :: _pendingConfirmations
                ConfirmationResult.NotConfirmedYet(confirmation)
            }
          }
        }
      }
    }
  }
}
